architecture.tex

```diff
@@ -53,58 +55,190 @@
-to be worth it for the advantages listed above.
-
-\subsection{The Core}
+rational literal type. Given a fraction of integers such as \texttt{1/2},
+Python will perform floating point division and produce
+\texttt{0.5}~\footnote{This is the behavior in Python 3. In Python 2,
+  \texttt{1/2} will perform integer division and produce \texttt{0}, unless
+  one uses \texttt{from \_\_future\_\_ import division}.}. Python uses eager
+evaluation, so expressions like \texttt{x + 1/2} will produce \texttt{x +
+  0.5}, and by the time any SymPy function sees the \texttt{1/2} it has
+already been converted to \texttt{0.5} by Python. However, for a CAS like
+SymPy, one typically wants to work with exact rational numbers whenever
+possible. Working around this is simple, however:  one can wrap one of the
+integers with \texttt{Integer}, like \texttt{x + Integer(1)/2}, or using
+\texttt{x + Rational(1, 2)}. SymPy provides a function \texttt{S} which can be
+used to convert objects to SymPy types with minimal typing, such as \texttt{x
+  + S(1)/2}. This gotcha is a small downside to using Python
+directly instead of a custom DSL, and we consider it to be worth it for the
+advantages listed above.
 
 %%
 %% Assumptions
 \subsection{Assumptions}
 
+An important feature of the SymPy core is the assumptions system. The
+assumptions system allows users to specify that symbols have certain common
+mathematical properties, such as being positive, imaginary, or integer. SymPy
+is careful to never perform simplifications on an expression unless the
+assumptions allow them. For instance, the identity $\sqrt{x^2} = x$ holds if
+$x$ is nonnegative ($x>=0$). If $x$ is real, the identity $\sqrt{x^2}=|x|$
+holds. However, for general complex $x$, no such identity holds.
+
+By default, SymPy performs all calculations assuming that variables are
+complex valued. This assumption makes it easier to treat mathematical problems
+in full generality.
+
+\begin{verbatim}
+>>> x = Symbol('x')
+>>> sqrt(x**2)
+sqrt(x**2)
+\end{verbatim}
+
+By assuming symbols are complex by default, SymPy avoids performing
+mathematically invalid operations. However, in many cases users will wish to
+simplify expressions containing terms like $\sqrt{x^2}$.
+
+Assumptions are set on \texttt{Symbol} objects when they are created. For
+instance \texttt{Symbol('x', positive=True)} will create a symbol named
+\texttt{x} that is assumed to be positive.
+
+\begin{verbatim}
+>>> x = Symbol('x', positive=True)
+>>> sqrt(x**2)
+x
+\end{verbatim}
+
+Some common assumptions that SymPy allows are \texttt{positive},
+\texttt{negative}, \texttt{real}, \texttt{nonpositive}, \texttt{nonnegative},
+\texttt{real}, \texttt{integer}, and \texttt{commutative}~\footnote{If $A$ and
+  $B$ are Symbols created with \texttt{commutative=False} then SymPy will keep
+  $AB$ and $BA$ distinct.}. Assumptions on any object can be checked with the
+\verb|is_|\texttt{\textit{assumption}} attributes, like \verb|x.is_positive|.
```

Comment:
Maybe emphasize that the assumption name has to be lowercase, mention difference between _is\_integer_ and _is\_Integer_.